module/zfs/dmu_send.c

```diff
@@ -4159,13 +4168,26 @@ add_ds_to_guidmap(const char *name, avl_tree_t *guid_map, uint64_t snapobj,
 	if (err != 0)
 		return (err);
 	gmep = kmem_alloc(sizeof (*gmep), KM_SLEEP);
-	err = dsl_dataset_hold_obj_flags(dp, snapobj, dsflags, gmep, &snapds);
+	err = dsl_dataset_own_obj(dp, snapobj, dsflags, gmep, &snapds);
 	if (err == 0) {
-		gmep->guid = dsl_dataset_phys(snapds)->ds_guid;
+		/*
+		 * If this is a deduplicated raw send stream, we need
+		 * to make sure that we can still read raw blocks from
+		 * earlier datasets in the stream, so we set the
+		 * os_raw_receive flag now.
+		 */
+		if (raw) {
+			if (dmu_objset_from_ds(snapds, &os)) {
```

Comment:
I'd suggest preserving and returning the real error.  It's currently unused by the caller so this doesn't really matter, but I don't see a compelling reason to convert it to `EINVAL`.